// react-lightning-design-system/lib/scripts/Card.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CardFooter = exports.CardBody = exports.CardHeader = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

var _Icon = require('./Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _Text = require('./Text');

var _Text2 = _interopRequireDefault(_Text);

var _MediaObject = require('./MediaObject');

var _MediaObject2 = _interopRequireDefault(_MediaObject);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Card = function Card(_ref) {
  var size = _ref.size,
      className = _ref.className,
      children = _ref.children;

  var cardClassNames = (0, _classnames2.default)(className, 'slds-card', {
    'slds-card--narrow': size === 'narrow'
  });
  var relativeContainerStyle = {
    position: 'relative'
  };
  return _react2.default.createElement(
    'article',
    { className: cardClassNames, style: relativeContainerStyle },
    children
  );
};

var CardHeader = exports.CardHeader = function CardHeader(_ref2) {
  var cardIconCategory = _ref2.cardIconCategory,
      cardIcon = _ref2.cardIcon,
      closeButton = _ref2.closeButton,
      children = _ref2.children;

  var icon = _react2.default.createElement(_Icon2.default, { category: cardIconCategory, icon: cardIcon, textColor: null });

  var closeButtonPosition = {
    position: 'absolute',
    top: '5px',
    right: '5px'
  };
  return _react2.default.createElement(
    'div',
    { className: 'slds-card__header slds-grid' },
    _react2.default.createElement(
      _MediaObject2.default,
      { className: 'slds-media--center slds-has-flexi-truncate', figureLeft: icon },
      _react2.default.createElement(
        _Text2.default,
        { tag: 'span', category: 'heading', type: 'small' },
        children
      )
    ),
    closeButton ? _react2.default.createElement(
      'div',
      { className: 'slds-no-flex', style: closeButtonPosition },
      _react2.default.createElement(_Button2.default, { type: 'icon-bare', icon: 'close', iconSize: 'medium' })
    ) : null
  );
};

CardHeader.propTypes = {
  cardIconCategory: _react.PropTypes.string,
  cardIcon: _react.PropTypes.string,
  closeButton: _react.PropTypes.bool,
  children: _react.PropTypes.node
};

var CardBody = exports.CardBody = function CardBody(_ref3) {
  var textAlign = _ref3.textAlign,
      children = _ref3.children;

  var styles = { textAlign: textAlign };
  return _react2.default.createElement(
    'div',
    { style: styles, className: 'slds-card__body' },
    children
  );
};

CardBody.propTypes = {
  children: _react.PropTypes.node,
  textAlign: _react.PropTypes.string
};

var CardFooter = exports.CardFooter = function CardFooter(_ref4) {
  var children = _ref4.children;
  return _react2.default.createElement(
    'div',
    { className: 'slds-card__footer' },
    children
  );
};

CardFooter.propTypes = {
  children: _react.PropTypes.node
};

Card.Header = CardHeader;
Card.Body = CardBody;
Card.Footer = CardFooter;

var CARD_SIZES = ['narrow'];

Card.propTypes = {
  className: _react.PropTypes.string,
  children: _react.PropTypes.node,
  size: _react.PropTypes.oneOf(CARD_SIZES)
};

exports.default = Card;
