// react-lightning-design-system/lib/scripts/FileSelector.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDropzone = require('react-dropzone');

var _reactDropzone2 = _interopRequireDefault(_reactDropzone);

var _uuid = require('uuid');

var _uuid2 = _interopRequireDefault(_uuid);

var _FormElement = require('./FormElement');

var _FormElement2 = _interopRequireDefault(_FormElement);

var _Button = require('./Button');

var _Button2 = _interopRequireDefault(_Button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FileSelector = function (_Component) {
  (0, _inherits3.default)(FileSelector, _Component);

  function FileSelector(props) {
    (0, _classCallCheck3.default)(this, FileSelector);

    var _this = (0, _possibleConstructorReturn3.default)(this, (FileSelector.__proto__ || (0, _getPrototypeOf2.default)(FileSelector)).call(this, props));

    _this.state = {
      errorMsg: null
    };
    _this.onDrop = _this.onDrop.bind(_this);
    _this.onDropError = _this.onDropError.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(FileSelector, [{
    key: 'onDrop',
    value: function onDrop(file) {
      if (this.props.isSlave) {
        this.props.onError(file[0] ? null : this.props.errorMessage);
        return;
      }
      this.setState({
        errorMsg: !file[0] ? this.props.errorMessage : null
      });
    }
  }, {
    key: 'onDropError',
    value: function onDropError(error) {
      this.setState({
        errorMsg: error
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          _props$id = _props.id,
          id = _props$id === undefined ? 'input-' + (0, _uuid2.default)() : _props$id,
          label = _props.label,
          accept = _props.accept,
          multiple = _props.multiple,
          buttonText = _props.buttonText,
          errorMessage = _props.errorMessage,
          props = (0, _objectWithoutProperties3.default)(_props, ['id', 'label', 'accept', 'multiple', 'buttonText', 'errorMessage']);


      var error = this.state.errorMsg;

      if (label || error) {
        var formElemProps = {
          id: id,
          label: label,
          error: error
        };

        return _react2.default.createElement(
          _FormElement2.default,
          formElemProps,
          _react2.default.createElement(FileSelector, (0, _extends3.default)({
            id: id,
            accept: accept,
            multiple: multiple,
            buttonText: buttonText,
            errorMessage: errorMessage,
            isSlave: true,
            onError: this.onDropError
          }, props))
        );
      }

      return _react2.default.createElement(
        'div',
        { className: 'slds-file-selector slds-file-selector--files' },
        _react2.default.createElement(
          _reactDropzone2.default,
          {
            multiple: multiple,
            accept: accept,
            onDrop: this.onDrop,
            className: 'slds-file-selector__dropzone',
            activeClassName: 'slds-has-drag-over',
            rejectClassName: 'slds-has-drag-over',
            inputProps: { className: 'slds-file-selector__input' }
          },
          _react2.default.createElement(
            'label',
            { className: 'slds-file-selector__body', htmlFor: id },
            _react2.default.createElement(
              _Button2.default,
              { className: 'slds-file-selector__button', type: 'neutral', icon: 'upload', iconAlign: 'left' },
              buttonText
            ),
            _react2.default.createElement(
              'span',
              { className: 'slds-file-selector__text slds-medium-show' },
              'or Drop Files'
            )
          )
        )
      );
    }
  }]);
  return FileSelector;
}(_react.Component);

exports.default = FileSelector;


FileSelector.propTypes = {
  id: _react.PropTypes.string,
  label: _react.PropTypes.string,
  error: _FormElement2.default.propTypes.error,
  errorMessage: _react.PropTypes.string,
  accept: _react.PropTypes.string,
  multiple: _react.PropTypes.bool,
  buttonText: _react.PropTypes.string,
  isSlave: _react.PropTypes.bool,
  onError: _react.PropTypes.func
};
