// react-lightning-design-system/lib/scripts/Tab.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Tab = function Tab(_ref) {
  var className = _ref.className,
      active = _ref.active,
      children = _ref.children,
      props = (0, _objectWithoutProperties3.default)(_ref, ['className', 'active', 'children']);

  var pprops = (0, _util.cleanProps)(props, Tab.propTypes);
  delete pprops.title;
  delete pprops.eventKey;
  delete pprops.menuItems;
  delete pprops.menuIcon;

  var tabClassNames = (0, _classnames2.default)(className, 'slds-tabs__content', 'slds-' + (active ? 'show' : 'hide'));
  return _react2.default.createElement(
    'div',
    (0, _extends3.default)({ className: tabClassNames, role: 'tabpanel' }, pprops),
    children
  );
};

Tab.propTypes = {
  className: _react.PropTypes.string,
  active: _react.PropTypes.bool,
  title: _react.PropTypes.string,
  menu: _react.PropTypes.element,
  menuItems: _react.PropTypes.arrayOf(_react.PropTypes.element),
  menuIcon: _react.PropTypes.string,
  children: _react.PropTypes.node,
  eventKey: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.number])
};

exports.default = Tab;
